pkg/webhook/elasticquota/quota_topology.go

```diff
@@ -0,0 +1,664 @@
+	if invalid, err := extension.IsForbiddenModify(quota); invalid {
+		return err
+	}
+
+	if parentName, exist := quota.Labels[extension.LabelQuotaParent]; exist && parentName != extension.RootQuotaName {
+		parentInfo, find := qt.quotaInfoMap[parentName]
+		if !find {
+			err = fmt.Errorf("%v has parentName %v but not found parentInfo", quota.Name, parentName)
+			return err
+		}
+		if parentInfo.Name != parentName {
+			err = fmt.Errorf("%v parentInfo.Name != quota.parentName:[%v,%v]", quota.Name, parentInfo.Name, parentName)
+			return err
+		}
+	}
+
+	// min and max's each dimension should not have negative value
+	for key, val := range quota.Spec.Max {
+		if val.Value() < 0 {
+			return fmt.Errorf("%v quota.Spec.Max[:%v < 0 %v", quota.Name, key, val.Value())
+		}
+	}
+	for key, val := range quota.Spec.Min {
+		if val.Value() < 0 {
+			return fmt.Errorf("%v quota.Spec.Min[:%v < 0 %v", quota.Name, key, val.Value())
+		}
+	}
+
+	// minQuota <= maxQuota
+	valid := true
+	for key, val := range quota.Spec.Min {
+		if maxVal, exist := quota.Spec.Max[key]; exist {
+			if maxVal.Cmp(val) == -1 {
+				valid = false
+			}
+		} else {
+			valid = false
+		}
+		if !valid {
+			err = fmt.Errorf("%v min :%v > max,%v", quota.Name, quota.Spec.Min, quota.Spec.Max)
+			return err
+		}
+	}
+	return nil
+}
+
+func (qt *quotaTopology) fillDefaultQuotaInfo(quota *v1alpha1.ElasticQuota) (err error) {
+	if quota.Labels == nil {
+		quota.Labels = make(map[string]string)
+	}
+	if quota.Annotations == nil {
+		quota.Annotations = make(map[string]string)
+	}
+
+	// if not configure
+	if parentName, exist := quota.Labels[extension.LabelQuotaParent]; !exist || len(parentName) == 0 {
+		quota.Labels[extension.LabelQuotaParent] = extension.RootQuotaName
+		klog.V(5).Infof("fill quota root parent info: %v", quota.Name)
+	}
+	maxQuota, err := json.Marshal(&quota.Spec.Max)
+	if err != nil {
+		err = fmt.Errorf("fillDefaultQuotaInfo marshal quota max failed:%v", err)
+		return err
+	}
+	if sharedWeight, exist := quota.Annotations[extension.AnnotationSharedWeight]; !exist || len(sharedWeight) == 0 {
+		quota.Annotations[extension.AnnotationSharedWeight] = string(maxQuota)
+	}
+	return nil
+}
+
+func (qt *quotaTopology) validateQuota(quotaInfo *core.QuotaInfo) (err error) {
+	if quotaInfo.ParentName == extension.RootQuotaName && !quotaInfo.IsParent {
+		return nil
+	}
+
+	if !qt.checkSubAndParentGroupMaxQuotaKeySame(quotaInfo) {
+		return fmt.Errorf("%v checkSubAndParentmaxQuotaKeySame failed", quotaInfo.Name)
+	}
+
+	if !qt.checkMaxQuota(quotaInfo) {
+		return fmt.Errorf("%v checkMaxQuota failed", quotaInfo.Name)
+	}
+
+	if !qt.checkMinQuotaSum(quotaInfo) {
+		return fmt.Errorf("%v checkMinQuotaSum failed", quotaInfo.Name)
+	}
+
+	return nil
+}
+
+func (qt *quotaTopology) checkSubAndParentGroupMaxQuotaKeySame(quotaInfo *core.QuotaInfo) bool {
+	if quotaInfo.ParentName == extension.RootQuotaName {
+		return true
+	}
+
+	parentInfo, exist := qt.quotaInfoMap[quotaInfo.ParentName]
```

Comment:
you only check modified and its parent,  why not check modifed and its child?

Reply:
fixed